### ISSUE TITLE
ID Tokens are not encrypted. Have I misconfigured something?

### ISSUE THREAD

**Issue by Edmund-Loh** [panva/node-oidc-provider#721]
I have been trying to understand the features of oidc-provider and have been trying my configurations out with https://oidcdebugger.com/

I have a use case where I need the ID Token to be encrypted. With the authorisation code that I got, I used Postman to call the token endpoint of oidc-provider. The ID token returned was unencrypted, in the form of a JWS Token. I am guessing if encryption were successful, the ID Token should take the format of a JWT Token instead?

This is my configuration. Key values have been redacted. Is there any configuration which I have set wrongly?
```
const configuration = {
  // ... see the available options in Configuration options section
  formats: {
    AccessToken: 'jwt',
  },
  clients: [{
    client_id: process.env.CLIENT_ID,
    client_secret: process.env.CLIENT_SECRET,
    redirect_uris: [process.env.CLIENT_URI],
    scope: 'openid',
    grant_types: ['authorization_code'],
    response_types: ['code'],
    token_endpoint_auth_method: 'client_secret_post',
    // + other client properties
  }],
  findAccount: async function findAccount(ctx, sub, token) {
    return {
      accountId: sub,
      async claims(use, scope, claims, rejected){

        return {
          sub: sub, // it is essential to always return a sub claim
          userInfo: {
            Type: 'User',
            FullName: 'John Doe',
            HOLDER: 'YES',
          },
          amr: ['pwd','sms'],
        };
      }
    };
  },
  features: {
    introspection: { enabled: true },
    encryption: { enabled: true },
    userinfo: { enabled: false },
    jwtUserinfo: { enabled: false },
  },
  jwks: {
  keys: [
    {
      e: '...',
      n: '...',
      d: '...',
      p: '...',
      q: '...',
      dp: '...',
      dq: '...',
      qi: '...',
      kty: 'RSA',
      kid: 'tFyIVxwGW9pwA_bkZw2b8yozwD1RAvr-ygDSXZUWuZ4',
      use: 'sig'
    },
    {
      e: '...',
      n: '...',
      d: '...',
      p: '...',
      q: '...',
      dp: '...',
      dq: '...',
      qi: '...',
      kty: 'RSA',
      kid: 'O3jYgk8qfMduRLy1f9aXihsjZegoIXKY7hBq8lZzKYI',
      use: 'enc'
    }
  ]
},
  claims: {
    openid: ['sub','amr','userInfo'],
  },
  extraAccessTokenClaims: async function extraAccessTokenClaims(ctx, token) {
    return {
      authorization : {
        group1 : {},
        group2 : {},
        group3 : {},
      },
    };
  },
  issueRefreshToken : async function issueRefreshToken(ctx, client, code) {
    //return client.grantTypeAllowed('refresh_token') && code.scopes.has('offline_access');
    return true;
  },
  ttl: {
    AccessToken: 1 * 60 * 60, // 1 hour in seconds
    AuthorizationCode: 10 * 60, // 10 minutes in seconds
    IdToken: 1 * 60 * 60, // 1 hour in seconds
    DeviceCode: 10 * 60, // 10 minutes in seconds
    RefreshToken: 1 * 24 * 60 * 60, // 1 day in seconds
  },
};
```

**Comment by panva** [comment]
Your configuration says encryption is available (see your openid-configuration) but your client is not configured to get encrypted tokens, you need to set id_token_encrypted_response_alg and optionally also id_token_encrypted_response_enc.

The form it will take is as per specification, first ID Token is signed using the configured id_token_signed_response_alg and that is then encrypted. It's a Nested JWT.

**Comment by Edmund-Loh** [comment]
Thank you very much for your prompt reply. I have set the values for "id_token_encrypted_response_alg" and "id_token_encrypted_response_enc". I have also provided a public key as the value for client.jwk. I am able to receive an encrypted ID token as a response now.

I first tried using "RSA-OAEP-256" for "id_token_encrypted_response_alg", but got an error message indicating that it was not permitted. I have since switched to "RSA-OAEP" instead. I am guessing "RSA-OAEP-256" is currently not supported because it is not part of the OpenID specifications?
